performance regression on ARM: 1.20 apiserver has high cpu usage

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->


**What happened**:
Cluster was upgraded from 1.19.5 to 1.20.1 using kubeadm. After that, kube-apiserver cpu usage went from 50% to 300%.

**What you expected to happen**:
Expected that apiserver has the same cpu usage.

**How to reproduce it (as minimally and precisely as possible)**:
Update a cluster from 1.19.5 to 1.20.1.

**Anything else we need to know?**:
The cluster is running on sbc hardware, thats why the original 50% cpu usage, and thats why 300% cpu usage is noticeable and causes service disruption.

**Environment**:
- Kubernetes version (use `kubectl version`): (the apiserver is rolled back to 1.19.5)
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.1", GitCommit:"c4d752765b3bbac2237bf87cf0b1c2e307844666", GitTreeState:"clean", BuildDate:"2020-12-18T12:09:25Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.5", GitCommit:"e338cf2c6d297aa603b50ad3a301f761b4173aa6", GitTreeState:"clean", BuildDate:"2020-12-09T11:10:32Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/arm"}
```
- Cloud provider or hardware configuration:
bare metal on SBC hardware.
- OS (e.g: `cat /etc/os-release`):
```
# cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```


- Kernel (e.g. `uname -a`):
```
# uname -a
Linux k8s-master01 5.4.86 #1 SMP Wed Dec 30 17:58:36 UTC 2020 armv7l GNU/Linux
```
- Install tools:
kubeadm
- Network plugin and version (if this is a network-related bug):
kube-router
- Others:
Please check profiling of each apiserver.
[apiserver.tar.gz](https://github.com/kubernetes/kubernetes/files/5764262/apiserver.tar.gz)

Compiled the test code for i386 resulted in:
```
$ GOARCH=386 go test -c .
$ ./metrics.test 
I0106 20:42:07.027159    3754 sample_and_watermark_test.go:54] Expect about 16 warnings about time going backwards; this is fake time deliberately misbehaving.
E0106 20:42:07.030225    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.033393962 +0100 CET m=+0.008874424 to 2021-01-06 20:42:07.032952962 +0100 CET m=+0.008433424 for labelValues=[]string{}
E0106 20:42:07.032126    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.033533962 +0100 CET m=+0.009014424 to 2021-01-06 20:42:07.033351962 +0100 CET m=+0.008832424 for labelValues=[]string{}
E0106 20:42:07.035344    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.038272962 +0100 CET m=+0.013753424 to 2021-01-06 20:42:07.038072962 +0100 CET m=+0.013553424 for labelValues=[]string{}
E0106 20:42:07.037692    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.039277962 +0100 CET m=+0.014758424 to 2021-01-06 20:42:07.038939962 +0100 CET m=+0.014420424 for labelValues=[]string{}
E0106 20:42:07.038499    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.039277962 +0100 CET m=+0.014758424 to 2021-01-06 20:42:07.038528962 +0100 CET m=+0.014009424 for labelValues=[]string{}
E0106 20:42:07.057499    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.061735962 +0100 CET m=+0.037216424 to 2021-01-06 20:42:07.061250962 +0100 CET m=+0.036731424 for labelValues=[]string{}
E0106 20:42:07.057994    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.061735962 +0100 CET m=+0.037216424 to 2021-01-06 20:42:07.061291962 +0100 CET m=+0.036772424 for labelValues=[]string{}
E0106 20:42:07.058501    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.061735962 +0100 CET m=+0.037216424 to 2021-01-06 20:42:07.061199962 +0100 CET m=+0.036680424 for labelValues=[]string{}
E0106 20:42:07.059981    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.065417962 +0100 CET m=+0.040898424 to 2021-01-06 20:42:07.065346962 +0100 CET m=+0.040827424 for labelValues=[]string{}
E0106 20:42:07.060463    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.065417962 +0100 CET m=+0.040898424 to 2021-01-06 20:42:07.065202962 +0100 CET m=+0.040683424 for labelValues=[]string{}
E0106 20:42:07.068628    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.073341962 +0100 CET m=+0.048822424 to 2021-01-06 20:42:07.073274962 +0100 CET m=+0.048755424 for labelValues=[]string{}
E0106 20:42:07.069204    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.073341962 +0100 CET m=+0.048822424 to 2021-01-06 20:42:07.072921962 +0100 CET m=+0.048402424 for labelValues=[]string{}
E0106 20:42:07.072850    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.076513962 +0100 CET m=+0.051994424 to 2021-01-06 20:42:07.076202962 +0100 CET m=+0.051683424 for labelValues=[]string{}
E0106 20:42:07.079595    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.087156962 +0100 CET m=+0.062637424 to 2021-01-06 20:42:07.086812962 +0100 CET m=+0.062293424 for labelValues=[]string{}
E0106 20:42:07.084317    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.100137962 +0100 CET m=+0.075618424 to 2021-01-06 20:42:07.099639962 +0100 CET m=+0.075120424 for labelValues=[]string{}
E0106 20:42:07.085404    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.100857962 +0100 CET m=+0.076338424 to 2021-01-06 20:42:07.100804962 +0100 CET m=+0.076285424 for labelValues=[]string{}
E0106 20:42:07.088857    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.102475962 +0100 CET m=+0.077956424 to 2021-01-06 20:42:07.102438962 +0100 CET m=+0.077919424 for labelValues=[]string{}
E0106 20:42:07.092304    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.110067962 +0100 CET m=+0.085548424 to 2021-01-06 20:42:07.109700962 +0100 CET m=+0.085181424 for labelValues=[]string{}
E0106 20:42:07.102035    3754 sample_and_watermark.go:202] Time went backwards from 2021-01-06 20:42:07.126274962 +0100 CET m=+0.101755424 to 2021-01-06 20:42:07.125979962 +0100 CET m=+0.101460424 for labelValues=[]string{}
--- FAIL: TestSampler (0.08s)
    sample_and_watermark_test.go:55: t0=2021-01-06 20:42:07.027078962 +0100 CET m=+0.002559424
    sample_and_watermark_test.go:70: For i=0, ddt=1.581ms, t1=2021-01-06 20:42:07.030659962 +0100 CET m=+0.006140424, diff=3.581ms, dt=3.581ms, count=3
    sample_and_watermark_test.go:70: For i=1, ddt=1.387ms, t1=2021-01-06 20:42:07.032046962 +0100 CET m=+0.007527424, diff=4.968ms, dt=4.968ms, count=4
    sample_and_watermark_test.go:70: For i=2, ddt=1.347ms, t1=2021-01-06 20:42:07.033393962 +0100 CET m=+0.008874424, diff=6.315ms, dt=6.315ms, count=6
    sample_and_watermark_test.go:70: For i=3, ddt=-441µs, t1=2021-01-06 20:42:07.032952962 +0100 CET m=+0.008433424, diff=5.874ms, dt=6.315ms, count=6
    sample_and_watermark_test.go:70: For i=4, ddt=581µs, t1=2021-01-06 20:42:07.033533962 +0100 CET m=+0.009014424, diff=6.455ms, dt=6.455ms, count=6
    sample_and_watermark_test.go:70: For i=5, ddt=-182µs, t1=2021-01-06 20:42:07.033351962 +0100 CET m=+0.008832424, diff=6.273ms, dt=6.455ms, count=6
    sample_and_watermark_test.go:70: For i=6, ddt=1.925ms, t1=2021-01-06 20:42:07.035276962 +0100 CET m=+0.010757424, diff=8.198ms, dt=8.198ms, count=8
    sample_and_watermark_test.go:70: For i=7, ddt=2.04ms, t1=2021-01-06 20:42:07.037316962 +0100 CET m=+0.012797424, diff=10.238ms, dt=10.238ms, count=10
    sample_and_watermark_test.go:70: For i=8, ddt=956µs, t1=2021-01-06 20:42:07.038272962 +0100 CET m=+0.013753424, diff=11.194ms, dt=11.194ms, count=11
    sample_and_watermark_test.go:70: For i=9, ddt=-200µs, t1=2021-01-06 20:42:07.038072962 +0100 CET m=+0.013553424, diff=10.994ms, dt=11.194ms, count=11
    sample_and_watermark_test.go:70: For i=10, ddt=1.194ms, t1=2021-01-06 20:42:07.039266962 +0100 CET m=+0.014747424, diff=12.188ms, dt=12.188ms, count=12
    sample_and_watermark_test.go:70: For i=11, ddt=11µs, t1=2021-01-06 20:42:07.039277962 +0100 CET m=+0.014758424, diff=12.199ms, dt=12.199ms, count=12
    sample_and_watermark_test.go:70: For i=12, ddt=-338µs, t1=2021-01-06 20:42:07.038939962 +0100 CET m=+0.014420424, diff=11.861ms, dt=12.199ms, count=12
    sample_and_watermark_test.go:70: For i=13, ddt=-411µs, t1=2021-01-06 20:42:07.038528962 +0100 CET m=+0.014009424, diff=11.45ms, dt=12.199ms, count=12
    sample_and_watermark_test.go:70: For i=14, ddt=2.228ms, t1=2021-01-06 20:42:07.040756962 +0100 CET m=+0.016237424, diff=13.678ms, dt=13.678ms, count=13
    sample_and_watermark_test.go:70: For i=15, ddt=774µs, t1=2021-01-06 20:42:07.041530962 +0100 CET m=+0.017011424, diff=14.452ms, dt=14.452ms, count=14
    sample_and_watermark_test.go:70: For i=16, ddt=1.711ms, t1=2021-01-06 20:42:07.043241962 +0100 CET m=+0.018722424, diff=16.163ms, dt=16.163ms, count=16
    sample_and_watermark_test.go:70: For i=17, ddt=945µs, t1=2021-01-06 20:42:07.044186962 +0100 CET m=+0.019667424, diff=17.108ms, dt=17.108ms, count=17
    sample_and_watermark_test.go:70: For i=18, ddt=1.737ms, t1=2021-01-06 20:42:07.045923962 +0100 CET m=+0.021404424, diff=18.845ms, dt=18.845ms, count=18
    sample_and_watermark_test.go:70: For i=19, ddt=606µs, t1=2021-01-06 20:42:07.046529962 +0100 CET m=+0.022010424, diff=19.451ms, dt=19.451ms, count=19
    sample_and_watermark_test.go:70: For i=20, ddt=1.995ms, t1=2021-01-06 20:42:07.048524962 +0100 CET m=+0.024005424, diff=21.446ms, dt=21.446ms, count=21
    sample_and_watermark_test.go:70: For i=21, ddt=1.966ms, t1=2021-01-06 20:42:07.050490962 +0100 CET m=+0.025971424, diff=23.412ms, dt=23.412ms, count=23
    sample_and_watermark_test.go:70: For i=22, ddt=2.028ms, t1=2021-01-06 20:42:07.052518962 +0100 CET m=+0.027999424, diff=25.44ms, dt=25.44ms, count=25
    sample_and_watermark_test.go:70: For i=23, ddt=758µs, t1=2021-01-06 20:42:07.053276962 +0100 CET m=+0.028757424, diff=26.198ms, dt=26.198ms, count=26
    sample_and_watermark_test.go:70: For i=24, ddt=547µs, t1=2021-01-06 20:42:07.053823962 +0100 CET m=+0.029304424, diff=26.745ms, dt=26.745ms, count=26
    sample_and_watermark_test.go:70: For i=25, ddt=2.447ms, t1=2021-01-06 20:42:07.056270962 +0100 CET m=+0.031751424, diff=29.192ms, dt=29.192ms, count=29
    sample_and_watermark_test.go:70: For i=26, ddt=1.787ms, t1=2021-01-06 20:42:07.058057962 +0100 CET m=+0.033538424, diff=30.979ms, dt=30.979ms, count=30
    sample_and_watermark_test.go:70: For i=27, ddt=1.388ms, t1=2021-01-06 20:42:07.059445962 +0100 CET m=+0.034926424, diff=32.367ms, dt=32.367ms, count=32
    sample_and_watermark_test.go:70: For i=28, ddt=2.29ms, t1=2021-01-06 20:42:07.061735962 +0100 CET m=+0.037216424, diff=34.657ms, dt=34.657ms, count=34
    sample_and_watermark_test.go:70: For i=29, ddt=-485µs, t1=2021-01-06 20:42:07.061250962 +0100 CET m=+0.036731424, diff=34.172ms, dt=34.657ms, count=34
    sample_and_watermark_test.go:70: For i=30, ddt=41µs, t1=2021-01-06 20:42:07.061291962 +0100 CET m=+0.036772424, diff=34.213ms, dt=34.657ms, count=34
    sample_and_watermark_test.go:70: For i=31, ddt=-92µs, t1=2021-01-06 20:42:07.061199962 +0100 CET m=+0.036680424, diff=34.121ms, dt=34.657ms, count=34
    sample_and_watermark_test.go:70: For i=32, ddt=1.887ms, t1=2021-01-06 20:42:07.063086962 +0100 CET m=+0.038567424, diff=36.008ms, dt=36.008ms, count=36
    sample_and_watermark_test.go:70: For i=33, ddt=2.331ms, t1=2021-01-06 20:42:07.065417962 +0100 CET m=+0.040898424, diff=38.339ms, dt=38.339ms, count=38
    sample_and_watermark_test.go:70: For i=34, ddt=-71µs, t1=2021-01-06 20:42:07.065346962 +0100 CET m=+0.040827424, diff=38.268ms, dt=38.339ms, count=38
    sample_and_watermark_test.go:70: For i=35, ddt=-144µs, t1=2021-01-06 20:42:07.065202962 +0100 CET m=+0.040683424, diff=38.124ms, dt=38.339ms, count=38
    sample_and_watermark_test.go:70: For i=36, ddt=1.237ms, t1=2021-01-06 20:42:07.066439962 +0100 CET m=+0.041920424, diff=39.361ms, dt=39.361ms, count=39
    sample_and_watermark_test.go:70: For i=37, ddt=131µs, t1=2021-01-06 20:42:07.066570962 +0100 CET m=+0.042051424, diff=39.492ms, dt=39.492ms, count=39
    sample_and_watermark_test.go:70: For i=38, ddt=985µs, t1=2021-01-06 20:42:07.067555962 +0100 CET m=+0.043036424, diff=40.477ms, dt=40.477ms, count=40
    sample_and_watermark_test.go:70: For i=39, ddt=526µs, t1=2021-01-06 20:42:07.068081962 +0100 CET m=+0.043562424, diff=41.003ms, dt=41.003ms, count=41
    sample_and_watermark_test.go:70: For i=40, ddt=1.913ms, t1=2021-01-06 20:42:07.069994962 +0100 CET m=+0.045475424, diff=42.916ms, dt=42.916ms, count=42
    sample_and_watermark_test.go:70: For i=41, ddt=590µs, t1=2021-01-06 20:42:07.070584962 +0100 CET m=+0.046065424, diff=43.506ms, dt=43.506ms, count=43
    sample_and_watermark_test.go:70: For i=42, ddt=1.694ms, t1=2021-01-06 20:42:07.072278962 +0100 CET m=+0.047759424, diff=45.2ms, dt=45.2ms, count=45
    sample_and_watermark_test.go:70: For i=43, ddt=1.063ms, t1=2021-01-06 20:42:07.073341962 +0100 CET m=+0.048822424, diff=46.263ms, dt=46.263ms, count=46
    sample_and_watermark_test.go:70: For i=44, ddt=-67µs, t1=2021-01-06 20:42:07.073274962 +0100 CET m=+0.048755424, diff=46.196ms, dt=46.263ms, count=46
    sample_and_watermark_test.go:70: For i=45, ddt=-353µs, t1=2021-01-06 20:42:07.072921962 +0100 CET m=+0.048402424, diff=45.843ms, dt=46.263ms, count=46
    sample_and_watermark_test.go:70: For i=46, ddt=578µs, t1=2021-01-06 20:42:07.073499962 +0100 CET m=+0.048980424, diff=46.421ms, dt=46.421ms, count=46
    sample_and_watermark_test.go:70: For i=47, ddt=824µs, t1=2021-01-06 20:42:07.074323962 +0100 CET m=+0.049804424, diff=47.245ms, dt=47.245ms, count=47
    sample_and_watermark_test.go:70: For i=48, ddt=659µs, t1=2021-01-06 20:42:07.074982962 +0100 CET m=+0.050463424, diff=47.904ms, dt=47.904ms, count=47
    sample_and_watermark_test.go:70: For i=49, ddt=853µs, t1=2021-01-06 20:42:07.075835962 +0100 CET m=+0.051316424, diff=48.757ms, dt=48.757ms, count=48
    sample_and_watermark_test.go:70: For i=50, ddt=457µs, t1=2021-01-06 20:42:07.076292962 +0100 CET m=+0.051773424, diff=49.214ms, dt=49.214ms, count=49
    sample_and_watermark_test.go:70: For i=51, ddt=221µs, t1=2021-01-06 20:42:07.076513962 +0100 CET m=+0.051994424, diff=49.435ms, dt=49.435ms, count=49
    sample_and_watermark_test.go:70: For i=52, ddt=-311µs, t1=2021-01-06 20:42:07.076202962 +0100 CET m=+0.051683424, diff=49.124ms, dt=49.435ms, count=49
    sample_and_watermark_test.go:70: For i=53, ddt=1.699ms, t1=2021-01-06 20:42:07.077901962 +0100 CET m=+0.053382424, diff=50.823ms, dt=50.823ms, count=50
    sample_and_watermark_test.go:70: For i=54, ddt=500µs, t1=2021-01-06 20:42:07.078401962 +0100 CET m=+0.053882424, diff=51.323ms, dt=51.323ms, count=51
    sample_and_watermark_test.go:70: For i=55, ddt=1.205ms, t1=2021-01-06 20:42:07.079606962 +0100 CET m=+0.055087424, diff=52.528ms, dt=52.528ms, count=52
    sample_and_watermark_test.go:70: For i=56, ddt=2.388ms, t1=2021-01-06 20:42:07.081994962 +0100 CET m=+0.057475424, diff=54.916ms, dt=54.916ms, count=54
    sample_and_watermark_test.go:70: For i=57, ddt=2.038ms, t1=2021-01-06 20:42:07.084032962 +0100 CET m=+0.059513424, diff=56.954ms, dt=56.954ms, count=56
    sample_and_watermark_test.go:70: For i=58, ddt=203µs, t1=2021-01-06 20:42:07.084235962 +0100 CET m=+0.059716424, diff=57.157ms, dt=57.157ms, count=57
    sample_and_watermark_test.go:70: For i=59, ddt=855µs, t1=2021-01-06 20:42:07.085090962 +0100 CET m=+0.060571424, diff=58.012ms, dt=58.012ms, count=58
    sample_and_watermark_test.go:70: For i=60, ddt=951µs, t1=2021-01-06 20:42:07.086041962 +0100 CET m=+0.061522424, diff=58.963ms, dt=58.963ms, count=58
    sample_and_watermark_test.go:70: For i=61, ddt=10µs, t1=2021-01-06 20:42:07.086051962 +0100 CET m=+0.061532424, diff=58.973ms, dt=58.973ms, count=58
    sample_and_watermark_test.go:70: For i=62, ddt=1.105ms, t1=2021-01-06 20:42:07.087156962 +0100 CET m=+0.062637424, diff=60.078ms, dt=60.078ms, count=60
    sample_and_watermark_test.go:70: For i=63, ddt=-344µs, t1=2021-01-06 20:42:07.086812962 +0100 CET m=+0.062293424, diff=59.734ms, dt=60.078ms, count=60
    sample_and_watermark_test.go:70: For i=64, ddt=1.766ms, t1=2021-01-06 20:42:07.088578962 +0100 CET m=+0.064059424, diff=61.5ms, dt=61.5ms, count=61
    sample_and_watermark_test.go:70: For i=65, ddt=2.328ms, t1=2021-01-06 20:42:07.090906962 +0100 CET m=+0.066387424, diff=63.828ms, dt=63.828ms, count=63
    sample_and_watermark_test.go:70: For i=66, ddt=2.061ms, t1=2021-01-06 20:42:07.092967962 +0100 CET m=+0.068448424, diff=65.889ms, dt=65.889ms, count=65
    sample_and_watermark_test.go:70: For i=67, ddt=702µs, t1=2021-01-06 20:42:07.093669962 +0100 CET m=+0.069150424, diff=66.591ms, dt=66.591ms, count=66
    sample_and_watermark_test.go:70: For i=68, ddt=2.283ms, t1=2021-01-06 20:42:07.095952962 +0100 CET m=+0.071433424, diff=68.874ms, dt=68.874ms, count=68
    sample_and_watermark_test.go:70: For i=69, ddt=246µs, t1=2021-01-06 20:42:07.096198962 +0100 CET m=+0.071679424, diff=69.12ms, dt=69.12ms, count=69
    sample_and_watermark_test.go:70: For i=70, ddt=2.063ms, t1=2021-01-06 20:42:07.098261962 +0100 CET m=+0.073742424, diff=71.183ms, dt=71.183ms, count=71
    sample_and_watermark_test.go:70: For i=71, ddt=1.876ms, t1=2021-01-06 20:42:07.100137962 +0100 CET m=+0.075618424, diff=73.059ms, dt=73.059ms, count=73
    sample_and_watermark_test.go:70: For i=72, ddt=-498µs, t1=2021-01-06 20:42:07.099639962 +0100 CET m=+0.075120424, diff=72.561ms, dt=73.059ms, count=73
    sample_and_watermark_test.go:70: For i=73, ddt=1.218ms, t1=2021-01-06 20:42:07.100857962 +0100 CET m=+0.076338424, diff=73.779ms, dt=73.779ms, count=73
    sample_and_watermark_test.go:70: For i=74, ddt=-53µs, t1=2021-01-06 20:42:07.100804962 +0100 CET m=+0.076285424, diff=73.726ms, dt=73.779ms, count=73
    sample_and_watermark_test.go:70: For i=75, ddt=1.594ms, t1=2021-01-06 20:42:07.102398962 +0100 CET m=+0.077879424, diff=75.32ms, dt=75.32ms, count=73
    sample_and_watermark_test.go:72: For i=75, t0=2021-01-06 20:42:07.027078962 +0100 CET m=+0.002559424, ddt=1.594ms, t1=2021-01-06 20:42:07.102398962 +0100 CET m=+0.077879424, expectedCount=75, actualCount=73
    sample_and_watermark_test.go:70: For i=76, ddt=77µs, t1=2021-01-06 20:42:07.102475962 +0100 CET m=+0.077956424, diff=75.397ms, dt=75.397ms, count=75
    sample_and_watermark_test.go:70: For i=77, ddt=-37µs, t1=2021-01-06 20:42:07.102438962 +0100 CET m=+0.077919424, diff=75.36ms, dt=75.397ms, count=75
    sample_and_watermark_test.go:70: For i=78, ddt=1.496ms, t1=2021-01-06 20:42:07.103934962 +0100 CET m=+0.079415424, diff=76.856ms, dt=76.856ms, count=76
    sample_and_watermark_test.go:70: For i=79, ddt=1.92ms, t1=2021-01-06 20:42:07.105854962 +0100 CET m=+0.081335424, diff=78.776ms, dt=78.776ms, count=78
    sample_and_watermark_test.go:70: For i=80, ddt=1.123ms, t1=2021-01-06 20:42:07.106977962 +0100 CET m=+0.082458424, diff=79.899ms, dt=79.899ms, count=79
    sample_and_watermark_test.go:70: For i=81, ddt=1.453ms, t1=2021-01-06 20:42:07.108430962 +0100 CET m=+0.083911424, diff=81.352ms, dt=81.352ms, count=81
    sample_and_watermark_test.go:70: For i=82, ddt=1.637ms, t1=2021-01-06 20:42:07.110067962 +0100 CET m=+0.085548424, diff=82.989ms, dt=82.989ms, count=82
    sample_and_watermark_test.go:70: For i=83, ddt=-367µs, t1=2021-01-06 20:42:07.109700962 +0100 CET m=+0.085181424, diff=82.622ms, dt=82.989ms, count=82
    sample_and_watermark_test.go:70: For i=84, ddt=741µs, t1=2021-01-06 20:42:07.110441962 +0100 CET m=+0.085922424, diff=83.363ms, dt=83.363ms, count=83
    sample_and_watermark_test.go:70: For i=85, ddt=1.559ms, t1=2021-01-06 20:42:07.112000962 +0100 CET m=+0.087481424, diff=84.922ms, dt=84.922ms, count=84
    sample_and_watermark_test.go:70: For i=86, ddt=1.533ms, t1=2021-01-06 20:42:07.113533962 +0100 CET m=+0.089014424, diff=86.455ms, dt=86.455ms, count=86
    sample_and_watermark_test.go:70: For i=87, ddt=1.143ms, t1=2021-01-06 20:42:07.114676962 +0100 CET m=+0.090157424, diff=87.598ms, dt=87.598ms, count=87
    sample_and_watermark_test.go:70: For i=88, ddt=2.391ms, t1=2021-01-06 20:42:07.117067962 +0100 CET m=+0.092548424, diff=89.989ms, dt=89.989ms, count=89
    sample_and_watermark_test.go:70: For i=89, ddt=1.502ms, t1=2021-01-06 20:42:07.118569962 +0100 CET m=+0.094050424, diff=91.491ms, dt=91.491ms, count=91
    sample_and_watermark_test.go:70: For i=90, ddt=378µs, t1=2021-01-06 20:42:07.118947962 +0100 CET m=+0.094428424, diff=91.869ms, dt=91.869ms, count=91
    sample_and_watermark_test.go:70: For i=91, ddt=836µs, t1=2021-01-06 20:42:07.119783962 +0100 CET m=+0.095264424, diff=92.705ms, dt=92.705ms, count=92
    sample_and_watermark_test.go:70: For i=92, ddt=46µs, t1=2021-01-06 20:42:07.119829962 +0100 CET m=+0.095310424, diff=92.751ms, dt=92.751ms, count=92
    sample_and_watermark_test.go:70: For i=93, ddt=1.607ms, t1=2021-01-06 20:42:07.121436962 +0100 CET m=+0.096917424, diff=94.358ms, dt=94.358ms, count=94
    sample_and_watermark_test.go:70: For i=94, ddt=2.44ms, t1=2021-01-06 20:42:07.123876962 +0100 CET m=+0.099357424, diff=96.798ms, dt=96.798ms, count=96
    sample_and_watermark_test.go:70: For i=95, ddt=2.003ms, t1=2021-01-06 20:42:07.125879962 +0100 CET m=+0.101360424, diff=98.801ms, dt=98.801ms, count=98
    sample_and_watermark_test.go:70: For i=96, ddt=52µs, t1=2021-01-06 20:42:07.125931962 +0100 CET m=+0.101412424, diff=98.853ms, dt=98.853ms, count=98
    sample_and_watermark_test.go:70: For i=97, ddt=343µs, t1=2021-01-06 20:42:07.126274962 +0100 CET m=+0.101755424, diff=99.196ms, dt=99.196ms, count=99
    sample_and_watermark_test.go:70: For i=98, ddt=-295µs, t1=2021-01-06 20:42:07.125979962 +0100 CET m=+0.101460424, diff=98.901ms, dt=99.196ms, count=99
    sample_and_watermark_test.go:70: For i=99, ddt=1.098ms, t1=2021-01-06 20:42:07.127077962 +0100 CET m=+0.102558424, diff=99.999ms, dt=99.999ms, count=99
FAIL
```

On i386 the test code triggers the bug too.

--- a/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark.go
+++ b/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark.go
@@ -160,17 +160,14 @@ func (saw *sampleAndWaterMarkHistograms) SetX1(x1 float64) {
 }
 
 func (saw *sampleAndWaterMarkHistograms) innerSet(updateXOrX1 func()) {
-	var when time.Time
-	var whenInt int64
-	var acc sampleAndWaterMarkAccumulator
-	var wellOrdered bool
-	func() {
+	when, whenInt, acc, wellOrdered := func() (time.Time, int64, sampleAndWaterMarkAccumulator, bool) {
 		saw.Lock()
 		defer saw.Unlock()
-		when = saw.clock.Now()
-		whenInt = saw.quantize(when)
-		acc = saw.sampleAndWaterMarkAccumulator
-		wellOrdered = !when.Before(acc.lastSet)
+		// Moved these variables here to tiptoe around https://github.com/golang/go/issues/43570 for #97685
+		when := saw.clock.Now()
+		whenInt := saw.quantize(when)
+		acc := saw.sampleAndWaterMarkAccumulator
+		wellOrdered := !when.Before(acc.lastSet)
 		updateXOrX1()
 		saw.relX = saw.x / saw.x1
 		if wellOrdered {
@@ -195,6 +192,7 @@ func (saw *sampleAndWaterMarkHistograms) innerSet(updateXOrX1 func()) {
 		} else if saw.relX > saw.hiRelX {
 			saw.hiRelX = saw.relX
 		}
+		return when, whenInt, acc, wellOrdered
 	}()
 	if !wellOrdered {
 		lastSetS := acc.lastSet.String()

